File: frontend/react/src/domain/incidentStatus.js
const TERMINAL_INCIDENT_STATUSES = new Set(["closed", "resolved", "failed", "cancelled", "canceled"]);
const PENDING_APPROVAL_STATUSES = new Set([
  "awaiting_approval",
  "pending_approval",
  "pending",
  "queued",
  "draft",
  "standby",
  "required",
]);

function normalizeStatus(value) {
  return String(value || "").trim().toLowerCase().replaceAll("-", "_").replaceAll(" ", "_");
}

const USER_LIFECYCLE_STAGES = Object.freeze({
  triage: new Set(["received", "normalized", "correlated", "open", "triaging"]),
  investigate: new Set(["context_ready", "evidence_pending", "evidence_ready", "rca_ready", "investigating", "needs_evidence"]),
  decide: new Set(["plan_ready", "policy_checked", "awaiting_approval", "pending_approval", "approved", "denied"]),
  act: new Set(["preflight", "dispatching", "executor_accepted", "executing", "running", "rolling_back"]),
  verify: new Set(["validating", "verifying", "resolved", "closed", "failed", "rolled_back", "manual_recovery_required"]),
});

export function userLifecycleStage(value) {
  const status = normalizeStatus(value);
  for (const [stage, statuses] of Object.entries(USER_LIFECYCLE_STAGES)) {
    if (statuses.has(status)) return stage;
  }
  return "triage";
}

export function decisionReadiness(input = {}) {
  const missing = [];
  if (Number(input.citationCoverage || 0) <= 0) missing.push("supporting citations");
  if (Number(input.evidenceCoverage || 0) < 0.6) missing.push("sufficient evidence coverage");
  if (input.evidenceFresh === false) missing.push("fresh evidence");
  if (Number(input.conflictCount || 0) > 0) missing.push("resolved conflicting evidence");
  if (!input.runbookAvailable) missing.push("an approved runbook");
  if (!input.preflightReady) missing.push("preflight readiness");
  if (!input.rollbackAvailable) missing.push("rollback readiness");
  if (missing.length) return { state: "not_ready", label: "Not ready—collect evidence", eligible: false, missing };
  if (input.dryRunComplete && input.risk !== "high" && input.risk !== "critical") {
    return { state: "execution_eligible", label: "Eligible for execution approval", eligible: true, missing: [] };
  }
  return { state: "dry_run_eligible", label: "Eligible for dry-run approval", eligible: true, missing: [] };
}

export function incidentStatusLabel(value) {
  const status = normalizeStatus(value);
  if (!status) return "Open";
  const labels = {
    awaiting_approval: "Awaiting approval",
    pending_approval: "Awaiting approval",
    policy_blocked: "Waiting for approval",
    dispatching: "Starting remediation",
    executor_accepted: "Queued for execution",
    execution_failed: "Execution failed",
    validation_failed: "Recovery validation failed",
    manual_intervention_required: "Manual intervention required",
  };
  return labels[status] || status.replaceAll("_", " ").replace(/^./, (character) => character.toUpperCase());
}

// Approval is a gate on execution. A stale or optimistic remediation status
// must never make the incident look as though a live action has started while
// the human decision is still pending.
export function effectiveIncidentStatus(incidentStatus, approvalStatus) {
  const incident = normalizeStatus(incidentStatus) || "open";
  const approval = normalizeStatus(approvalStatus);
  if (!TERMINAL_INCIDENT_STATUSES.has(incident) && PENDING_APPROVAL_STATUSES.has(approval)) {
    return "awaiting_approval";
  }
  return incident;
}

// A Jenkins queue URL is positive evidence that the latest submission is still
// active. It must take precedence over a stale incident failure left by an
// earlier attempt, while a confirmed successful outcome remains terminal.
export function effectiveExecutionStatus(incidentStatus, remediationStatus, queueUrl) {
  const incident = normalizeStatus(incidentStatus);
  const remediation = normalizeStatus(remediationStatus);
  // A policy block that routes the incident to a human is a decision gate,
  // not an execution failure. Keep the operator-facing lifecycle aligned with
  // the canonical incident state while retaining the raw action in technical
  // details for auditability.
  if (PENDING_APPROVAL_STATUSES.has(incident) && remediation === "policy_blocked") {
    return "awaiting_approval";
  }
  if (["succeeded", "completed", "closed", "resolved", "diagnostic_completed"].includes(remediation)
      || ["closed", "resolved"].includes(incident)) {
    return "succeeded";
  }
  if (["failed", "skipped", "policy_blocked", "dispatch_failed", "execution_failed", "validation_failed", "rollback_failed", "timed_out", "cancelled", "manual_intervention_required"].includes(remediation)) {
    return remediation;
  }
  if (["rolled_back", "dispatching", "executor_accepted", "running", "verifying", "rolling_back"].includes(remediation)) {
    return remediation;
  }
  if (String(queueUrl || "").trim() && !remediation) {
    return "executor_accepted";
  }
  if (incident === "failed") {
    return "failed";
  }
  return remediation;
}

const SUCCESS_EXECUTION_STATUSES = new Set(["succeeded", "completed", "closed", "resolved", "diagnostic_completed"]);
const FAILED_EXECUTION_STATUSES = new Set([
  "failed", "skipped", "policy_blocked", "dispatch_failed", "execution_failed", "validation_failed",
  "rollback_failed", "timed_out", "cancelled", "canceled", "manual_intervention_required",
]);
const ACTIVE_EXECUTION_STATUSES = new Set(["dispatching", "executor_accepted", "queued", "running", "verifying", "rolling_back"]);

// A single view model for every executor progress surface. Queue and build URLs
// remain useful links after a run finishes, but are not evidence that the run
// is still queued once the action has a terminal status.
export function executionProcessPresentation(executionStatus, dryRun, configured = false) {
  const status = normalizeStatus(executionStatus);
  const succeeded = SUCCESS_EXECUTION_STATUSES.has(status);
  const failed = FAILED_EXECUTION_STATUSES.has(status);
  const active = ACTIVE_EXECUTION_STATUSES.has(status);
  const queued = status === "executor_accepted" || status === "queued" || status === "dispatching";
  const running = active && !queued;
  const submitted = succeeded || failed || active || status === "rolled_back";
  const mode = dryRun == null
    ? submitted ? failed ? "Failed" : succeeded ? "Completed" : "Submitted" : "Not run"
    : String(dryRun).toLowerCase() === "true" ? "Validation" : "Live";
  return {
    status,
    succeeded,
    failed,
    active,
    submitted,
    badgeLabel: succeeded ? "Completed" : failed ? "Failed" : queued ? "Queued" : running ? "Running" : configured ? "Configured" : "Not configured",
    badgeClass: succeeded ? "pill-success" : failed ? "status-failed" : active ? "pill-info" : configured ? "pill-warning" : "status-failed",
    executionMode: mode,
    executionStageLabel: succeeded ? "Completed" : failed ? "Failed" : queued ? "Build queued" : running ? "Build running" : "Waiting",
    validationStageLabel: succeeded ? "Recovery validated" : failed ? "Blocked by execution failure" : "Waiting for result",
  };
}

File: backend/src/common/common/incident_status.py
from __future__ import annotations

from datetime import datetime, timezone
from typing import Any


SUCCESS_TERMINAL_STATUSES = frozenset({"closed", "resolved"})
TERMINAL_STATUSES = frozenset({*SUCCESS_TERMINAL_STATUSES, "cancelled", "canceled"})
ACTIVE_ACTION_STATUSES = frozenset(
    {"pending", "dispatching", "executor_accepted", "queued", "running", "verifying", "rolling_back"}
)
FAILED_ACTION_STATUSES = frozenset(
    {
        "failed",
        "rejected",
        "skipped",
        "dispatch_failed",
        "execution_failed",
        "validation_failed",
        "timed_out",
        "cancelled",
        "canceled",
        "rollback_failed",
        "manual_intervention_required",
    }
)
PENDING_APPROVAL_STATUSES = frozenset(
    {"awaiting_approval", "pending_approval", "pending", "queued", "draft", "standby", "required"}
)
APPROVED_STATUSES = frozenset({"approved", "approve", "modified", "modify"})


def _status(value: Any) -> str:
    return str(value or "").strip().lower().replace("-", "_").replace(" ", "_")


def _timestamp(value: Any) -> datetime:
    if isinstance(value, datetime):
        return value if value.tzinfo else value.replace(tzinfo=timezone.utc)
    if isinstance(value, str) and value.strip():
        try:
            parsed = datetime.fromisoformat(value.strip().replace("Z", "+00:00"))
            return parsed if parsed.tzinfo else parsed.replace(tzinfo=timezone.utc)
        except ValueError:
            pass
    return datetime.min.replace(tzinfo=timezone.utc)


def _action_incident_status(action_status: str) -> tuple[str, str] | None:
    if action_status in ACTIVE_ACTION_STATUSES:
        return "remediating", "The latest remediation action is active."
    if action_status in {"succeeded", "completed"}:
        return "validating", "Remediation completed and recovery validation is pending."
    if action_status in {"awaiting_approval", "policy_blocked"}:
        return "awaiting_approval", "Execution is waiting for an operator decision."
    if action_status in FAILED_ACTION_STATUSES:
        return "failed", "The latest remediation or recovery action failed."
    if action_status == "rolled_back":
        return "failed", "The remediation was rolled back and requires review."
    return None


def reduce_incident_status(
    *,
    projection_status: Any = None,
    projection_updated_at: Any = None,
    canonical_status: Any = None,
    canonical_updated_at: Any = None,
    approval_status: Any = None,
    approval_updated_at: Any = None,
    action_status: Any = None,
    action_updated_at: Any = None,
    closure_kind: Any = None,
) -> dict[str, str]:
    """Return one lifecycle status from the latest durable incident facts.

    Explicit approval and executor records are stronger than optimistic UI or
    projection state. Successful closure is monotonic. A newer approval may
    supersede an older policy block or failed attempt, while a newer failed
    execution must supersede a stale canonical ``approved`` value.
    """

    projection = _status(projection_status)
    canonical = _status(canonical_status)
    approval = _status(approval_status)
    action = _status(action_status)
    closure = _status(closure_kind)

    if canonical in SUCCESS_TERMINAL_STATUSES or projection in SUCCESS_TERMINAL_STATUSES:
        status = "closed" if "closed" in {canonical, projection} else "resolved"
        if closure == "manual":
            reason = "An authorized operator administratively closed the incident without a technical recovery claim."
        elif closure == "diagnostic":
            reason = "Diagnostic work was completed and closed without a technical recovery claim."
        else:
            reason = "Recovery validation closed the incident."
        return {"status": status, "source": "closure", "reason": reason}
    if canonical in {"cancelled", "canceled"} or projection in {"cancelled", "canceled"}:
        return {"status": "cancelled", "source": "incident", "reason": "The incident was explicitly cancelled."}

    approval_at = _timestamp(approval_updated_at)
    action_at = _timestamp(action_updated_at)
    canonical_at = _timestamp(canonical_updated_at)
    projection_at = _timestamp(projection_updated_at)

    action_fact = _action_incident_status(action)

    # A fresh analysis can deliberately reopen an incident after an older
    # approval or execution attempt. The materialized projection is the durable
    # result of that analysis; stale action/approval/canonical rows must not
    # keep the read model pinned to a superseded failure. Successful closure
    # remains monotonic because it is handled above.
    if (
        projection
        and projection_at > max(canonical_at, approval_at, action_at)
    ):
        return {
            "status": projection,
            "source": "projection",
            "reason": "A newer persisted lifecycle projection supersedes earlier approval or remediation attempts.",
        }

    if action_fact is not None and action_at >= approval_at:
        status, reason = action_fact
        return {"status": status, "source": "remediation_action", "reason": reason}

    if approval in APPROVED_STATUSES and approval_at >= action_at:
        return {
            "status": "approved",
            "source": "approval",
            "reason": "Approval is recorded; execution is the next required lifecycle action.",
        }
    if approval in PENDING_APPROVAL_STATUSES and approval_at >= action_at:
        return {
            "status": "awaiting_approval",
            "source": "approval",
            "reason": "The current remediation plan is waiting for approval.",
        }
    if approval in {"rejected", "reject", "denied"} and approval_at >= action_at:
        return {"status": "failed", "source": "approval", "reason": "Automation was rejected and requires escalation."}
    if action_fact is not None:
        status, reason = action_fact
        return {"status": status, "source": "remediation_action", "reason": reason}

    # A failed canonical record is stronger than a non-terminal projection, but
    # otherwise prefer whichever of the two materialized incident facts is newer.
    if canonical == "failed" and canonical_at >= projection_at:
        return {"status": "failed", "source": "incident", "reason": "The canonical incident lifecycle failed."}
    fallback = canonical if canonical_at >= projection_at else projection
    fallback_source = "incident" if canonical_at >= projection_at else "projection"
    fallback = fallback or canonical or projection or "open"
    return {"status": fallback, "source": fallback_source, "reason": "Derived from the latest persisted incident state."}
